### PR TITLE
Feature/cmaf ham hls

### PR DESCRIPTION
**Pull request by NoeliaBentancor**
## Description
Create m3u8 function so as to parse manifest hls to ham

## Requirements Checklist
- [ ] Unit Tests updated or fixed
- [ ] Docs/guides updated

### CODE REVIEW

File: lib/src/cmaf/ham/SwitchingSet.ts [hunk 1]
```diff
@@ -1,20 +1,21 @@
 import { Track } from "./Track";
 
+
 export class SwitchingSet {
-    
     id: string;
-    type: string;
+    type: string ;
```
Comment (mpollingerQualabs) [inline lib/src/cmaf/ham/SwitchingSet.ts]:
```suggestion
    type: string;
```

File: lib/src/cmaf/ham/SwitchingSet.ts [hunk 2]
```diff
@@ -1,20 +1,21 @@
 import { Track } from "./Track";
 
+
```
Comment (mpollingerQualabs) [inline lib/src/cmaf/ham/SwitchingSet.ts]:
```suggestion
```

File: lib/src/cmaf/ham/SwitchingSet.ts [hunk 3]
```diff
@@ -1,20 +1,21 @@
...
     id: string;
-    type: string;
+    type: string ;
     codec:string;
     duration:number;
     language:string;
     tracks : Track[];
 
-    constructor(id:string, type:string,codec:string,duration:number,language:string, tracks:Track[]) {
+    constructor(id:string, type:string ,codec:string,duration:number,language:string, tracks:Track[]) {
```
Comment (mpollingerQualabs) [inline lib/src/cmaf/ham/SwitchingSet.ts]:
```suggestion
    constructor(id:string, type:string, codec:string, duration:number, language:string, tracks:Track[]) {
```

File: lib/src/cmaf/ham/SwitchingSet.ts [hunk 4]
```diff
@@ -1,20 +1,21 @@
...
         this.id = id;
         this.type = type;
         this.codec = codec;
         this.duration = duration;
         this.language = language;
         this.tracks = tracks;
     }
+
```
Comment (mpollingerQualabs) [inline lib/src/cmaf/ham/SwitchingSet.ts]:
```suggestion
```

File: lib/src/cmaf/ham/Track.ts [hunk 5]
```diff
@@ -1,17 +1,24 @@
...
     id: string;
     type: string;
     codec:string;
     duration:number;
     language:string;
     bandwidth:number;
+    segments: Segment[];
 
-    constructor(id:string, type:string,codec:string,duration:number,language:string,bandwidth:number) {
+    constructor(id:string, type:string,codec:string,duration:number,language:string,bandwidth:number,segments: Segment[]) {
```
Comment (mpollingerQualabs) [inline lib/src/cmaf/ham/Track.ts]:
```suggestion
    constructor(id:string, type:string, codec:string, duration:number, language:string, bandwidth:number, segments: Segment[]) {
```

File: lib/src/cmaf/ham/VideoTrack.ts [hunk 6]
```diff
@@ -1,15 +1,21 @@
...
 export class VideoTrack extends Track{
     width:number;
     height:number;
     frameRate:number;
 
 
-    constructor(id:string, type:string,codec:string,duration:number,language:string,bandwidth:number, width:number, height:number,frameRate:number) {
-        super(id, type,codec,duration,language,bandwidth);
+    constructor(id:string, type:string,codec:string,duration:number,language:string,bandwidth:number, width:number, height:number,frameRate:number, segments: Segment[]) {
+        super(id, type,codec,duration,language,bandwidth,segments);
```
Comment (mpollingerQualabs) [inline lib/src/cmaf/ham/VideoTrack.ts]:
```suggestion
    constructor(id:string, type:string, codec:string, duration:number, language:string, bandwidth:number, width:number, height:number, frameRate:number, segments: Segment[]) {
        super(id, type, codec, duration, language, bandwidth, segments);
```

File: lib/src/cmaf/ham/manifestParser.ts [hunk 7]
```diff
@@ -7,8 +20,52 @@ async function readHLS(manifestUrl: string): Promise<string> {
...
+    const mediaGroupsAudio = parsedM3u8.mediaGroups?.AUDIO;
+    let audioSwitchingSets: SwitchingSet[] = [];
+    let audioTracks : AudioTrack[] = [];
+    let selectionSets: SelectionSet[] = [];
     
+    for (let audio in mediaGroupsAudio) {
+        for (let attributes in mediaGroupsAudio[audio]) {
+            let language = mediaGroupsAudio[audio][attributes].language;
+            audioSwitchingSets.push(new SwitchingSet(audio, AUDIO_TYPE, '', 0, language,[]));
+            audioTracks.push(new AudioTrack(audio, AUDIO_TYPE, '', 0, language, 0, 0, 0,[]));
```
Comment (mpollingerQualabs) [inline lib/src/cmaf/ham/manifestParser.ts]: Why some hardcoded values when creating the `SwitchingSet` and the `AudioTrack`?

**Review by mpollingerQualabs — COMMENTED** [review]
Added some suggestions for better readability. Should probably apply linter before commit

File: lib/src/cmaf/ham/manifestParser.ts (hunk 7 above)
Comment (NoeliaBentancor) [reply lib/src/cmaf/ham/manifestParser.ts]: I did not have that information. 
I will take a second look so as to check if i can fill them or find a better way of filling them.

File: lib/src/cmaf/ham/AudioTrack.ts [hunk 8]
```diff
@@ -1,11 +1,11 @@
...
     sampleRate:number;
     channels:number;
 
-    constructor(id:string, type:string,codec:string,duration:number,language:string,bandwidth:number, sampleRate:number, channels:number) {
-        super(id, type,codec,duration,language,bandwidth);
+    constructor(id:string, type:string,codec:string,duration:number,language:string,bandwidth:number, sampleRate:number, channels:number, segments:Segment[]) {
+        super(id, type,codec,duration,language,bandwidth,segments);
         this.sampleRate = sampleRate;
         this.channels = channels;
     }
```
Comment (DuckyCB) [inline lib/src/cmaf/ham/AudioTrack.ts]: Please run the linter, indentation should be tabs and a newline should be present at the end of the file

File: lib/src/cmaf/ham/manifestParser.ts [hunk 9]
```diff
@@ -7,8 +20,52 @@ async function readHLS(manifestUrl: string): Promise<string> {
...
+            if (audioSearched && !audioInSwitchingSet) switchingSets.push(audioSearched);
+        }
+
+        await Promise.all(parsedHlsManifest?.segments?.map(async (segment:any) => {
+            let {LANGUAGE, CODECS, BANDWIDTH } = playlist.attributes;
+            let {duration,uri} = segment;
+            let resolution = {width: playlist.attributes.RESOLUTION.width, height: playlist.attributes.RESOLUTION.height};
+            let { length, offset } = segment.byterange;
+            selectionSetDuration += duration;
+            let segments = [new Segment(duration ,uri,`${length}@${offset}`)];
```
Comment (DuckyCB) [inline lib/src/cmaf/ham/manifestParser.ts]: If the value is not reassigned, const should be used instead of let

Comment (NoeliaBentancor) [reply lib/src/cmaf/ham/manifestParser.ts]: Resolved

**Review by matiasrb97 — APPROVED** [review]

File: lib/src/cmaf/ham/Presentation.ts [hunk 10]
```diff
@@ -2,13 +2,15 @@ import { SelectionSet } from './SelectionSet.js';
 
 export class Presentation {
 	id: string;
-	duration: number;
-	selectionSets: SelectionSet[];
+	selectionsSets: SelectionSet[];
```
Comment (mpollingerQualabs) [inline lib/src/cmaf/ham/Presentation.ts]:
```suggestion
	selectionSets: SelectionSet[];
```

File: lib/src/cmaf/ham/Presentation.ts [hunk 11]
```diff
@@ -2,13 +2,15 @@ import { SelectionSet } from './SelectionSet.js';
 
 export class Presentation {
 	id: string;
-	duration: number;
-	selectionSets: SelectionSet[];
+	selectionsSets: SelectionSet[];
 
-	constructor(id: string, duration: number, selectionSet: SelectionSet[]) {
+	constructor(id: string,selectionsSet:SelectionSet[]) {
```
Comment (mpollingerQualabs) [inline lib/src/cmaf/ham/Presentation.ts]:
```suggestion
	constructor(id: string, selectionSets: SelectionSet[]) {
```

File: lib/src/cmaf/ham/Presentation.ts [hunk 12]
```diff
@@ -2,13 +2,15 @@ import { SelectionSet } from './SelectionSet.js';
...
-	duration: number;
-	selectionSets: SelectionSet[];
+	selectionsSets: SelectionSet[];
 
-	constructor(id: string, duration: number, selectionSet: SelectionSet[]) {
+	constructor(id: string,selectionsSet:SelectionSet[]) {
 		this.id = id;
-		this.duration = duration;
-		this.selectionSets = selectionSet;
+		this.selectionsSets = selectionsSet;
```
Comment (mpollingerQualabs) [inline lib/src/cmaf/ham/Presentation.ts]:
```suggestion
		this.selectionSets = selectionSets;
```

File: lib/src/cmaf/ham/SelectionSet.ts [hunk 13]
```diff
@@ -2,12 +2,10 @@ import { SwitchingSet } from './SwitchingSet.js';
 
 export class SelectionSet {
 	id: string;
-	duration: number;
-	switchingSet: SwitchingSet[];
+	switchingsSet:SwitchingSet[];
```
Comment (mpollingerQualabs) [inline lib/src/cmaf/ham/SelectionSet.ts]:
```suggestion
	switchingSets: SwitchingSet[];
```

File: lib/src/cmaf/ham/SelectionSet.ts [hunk 14]
```diff
@@ -2,12 +2,10 @@ import { SwitchingSet } from './SwitchingSet.js';
 
 export class SelectionSet {
 	id: string;
-	duration: number;
-	switchingSet: SwitchingSet[];
+	switchingsSet:SwitchingSet[];
 
-	constructor(id: string, duration: number, switchingSet: SwitchingSet[]) {
+	constructor(id: string, switchingsSet:SwitchingSet[]) {
```
Comment (mpollingerQualabs) [inline lib/src/cmaf/ham/SelectionSet.ts]:
```suggestion
	constructor(id: string, switchingSets: SwitchingSet[]) {
```

File: lib/src/cmaf/ham/SelectionSet.ts [hunk 15]
```diff
@@ -2,12 +2,10 @@ import { SwitchingSet } from './SwitchingSet.js';
...
-	duration: number;
-	switchingSet: SwitchingSet[];
+	switchingsSet:SwitchingSet[];
 
-	constructor(id: string, duration: number, switchingSet: SwitchingSet[]) {
+	constructor(id: string, switchingsSet:SwitchingSet[]) {
 		this.id = id;
-		this.duration = duration;
-		this.switchingSet = switchingSet;
+		this.switchingsSet = switchingsSet;
```
Comment (mpollingerQualabs) [inline lib/src/cmaf/ham/SelectionSet.ts]:
```suggestion
		this.switchingSets = switchingSets;
```

File: lib/src/cmaf/ham/Track.ts [hunk 16]
```diff
@@ -1,29 +1,19 @@
...
-	constructor(
-		id: string,
-		type: string,
-		codec: string,
-		duration: number,
-		language: string,
-		bandwidth: number,
-		segments: Segment[],
-	) {
+	constructor(id:string,  codec:string, duration:number, language:string, bandwidth:number, segments: Segment[]) {
```
Comment (mpollingerQualabs) [inline lib/src/cmaf/ham/Track.ts]:
```suggestion
	constructor(id: string, codec: string, duration: number, language: string, bandwidth: number, segments: Segment[]) {

```

**Review by mpollingerQualabs — CHANGES_REQUESTED** [review]

File: lib/src/cmaf/ham/Track.ts (hunk 16 above)
Comment (NoeliaBentancor) [reply lib/src/cmaf/ham/Track.ts]: This was changed when conflicts were resolved

File: lib/src/cmaf/ham/SelectionSet.ts (hunk 15 above)
Comment (NoeliaBentancor) [reply lib/src/cmaf/ham/SelectionSet.ts]: This was changed when conflicts were resolved

**Pull request merged**